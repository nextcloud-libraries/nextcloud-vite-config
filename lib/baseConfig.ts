/**
 * SPDX-FileCopyrightText: 2023 Ferdinand Thiessen <opensource@fthiessen.de>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { type CoreJSPluginOptions, corejsPlugin } from 'rollup-plugin-corejs'
import { minify as minifyPlugin } from 'rollup-plugin-esbuild-minify/lib/index.js'
import { nodePolyfills } from 'vite-plugin-node-polyfills'
import { defineConfig, mergeConfig, type UserConfigExport, type UserConfigFn } from 'vite'
import { RemoveEnsureWatchPlugin } from './plugins/RemoveEnsureWatch.js'

import replace from '@rollup/plugin-replace'
import vue2 from '@vitejs/plugin-vue2'
import browserslistToEsbuild from 'browserslist-to-esbuild'
import license from 'rollup-plugin-license'
import injectCSSPlugin from 'vite-plugin-css-injected-by-js'

type NodePolyfillsOptions = Parameters<typeof nodePolyfills>[0]

export interface BaseOptions {
	/** Strings to replace within your code */
	replace?: Record<string, string>
	/**
	 * Inject all styles inside the javascript bundle instead of emitting a .css file
	 * @default false
	 */
	inlineCSS?: boolean
	/**
	 * Whether to minify the output
	 * @default isProduction Minify on production, do not on development
	 */
	minify?: boolean
	/**
	 * Inject polyfills for node packages
	 * @default false Disabled to reduce buildtime
	 */
	nodePolyfills?: boolean | NodePolyfillsOptions
	/**
	 * Adjust settings for core-js polyfills
	 *
	 * By default enabled with `{ usage: true }`
	 */
	coreJS?: boolean | CoreJSPluginOptions
	/**
	 * Vite config to override or extend the base config
	 */
	config?: UserConfigExport
}

/**
 * Create a basic configuration
 * @param options Options to use
 */
export function createBaseConfig(options: BaseOptions = {}): UserConfigFn {
	return async (env) => {
		// get current build mode which is not necessary the same as the current `process.env`
		const { mode } = env
		const isDev = mode === 'development'
		// Set default values for optional options
		options = { minify: !isDev, replace: {}, config: {}, ...options }

		// This config is used to extend or override our base config
		// Make sure we get a user config and not a promise or a user config function
		const userConfig = await Promise.resolve(typeof options.config === 'function' ? options.config(env) : options.config)

		const plugins = []
		if (options?.inlineCSS !== false) {
		// Inject all imported styles into the javascript bundle
			plugins.push(injectCSSPlugin())
		}
		if (options?.nodePolyfills) {
		// Add polyfills for node packages
			plugins.push(nodePolyfills(typeof options.nodePolyfills === 'object' ? options.nodePolyfills : {}))
		}
		if (Object.keys(options.replace).length > 0) {
		// Replace global variables, built-in `define` option does not work (replaces also strings in 'node_modules/`)
			plugins.push(replace({
				preventAssignment: true,
				delimiters: ['\\b', '\\b'],
				include: ['src/**/*', 'lib/**/*', 'node_modules/@nextcloud/vue/**/*'],
				values: options.replace,
			}))
		}
		if (options.coreJS !== false) {
		// Add required polyfills, by default browserslist config is used
			plugins.push(corejsPlugin(typeof options.coreJS === 'object' ? options.coreJS : undefined))
		}

		return mergeConfig(defineConfig({
			plugins: [
				// Fix build in watch mode with commonjs files
				RemoveEnsureWatchPlugin,
				// Add vue2 support
				vue2({
					isProduction: !isDev,
					style: {
						trim: true,
					},
					template: {
						compilerOptions: {
							comments: isDev,
						},
					},
				}),
				// Add custom plugins
				...plugins,
				// Remove unneeded whitespace
				options?.minify ? minifyPlugin() : undefined,
				// Add license header with all dependencies
				license({
					sourcemap: true,
					banner: {
						commentStyle: 'regular',
						content: () => {
							// This should work with ESM, but esbuild converts it to a data URL which then will not work
							// const template = new URL('../banner-template.txt', import.meta.url);
							const template = resolve(import.meta.url.split('/').slice(1, -2).join('/'), 'banner-template.txt')
							return readFileSync(template, 'utf-8')
						},
					},
				}),
			],
			define: {
				// process env replacement (keep order of this rules)
				'globalThis.process.env.NODE_ENV': mode,
				'globalThis.process.env.': '({}).',
				'global.process.env.NODE_ENV': mode,
				'global.process.env.': '({}).',
				'process.env.NODE_ENV': mode,
				'process.env.': '({}).',
			},
			esbuild: {
				legalComments: 'inline',
				target: browserslistToEsbuild(),
			},
			build: {
				cssTarget: browserslistToEsbuild(),
				lib: {
					formats: ['es'],
					entry: {},
				},
				sourcemap: isDev || 'hidden',
				target: browserslistToEsbuild(),
			},
		}),
		// Add overrides from user config
		userConfig)
	}
}
