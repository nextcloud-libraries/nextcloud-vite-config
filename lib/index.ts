/**
 * SPDX-FileCopyrightText: 2023 Ferdinand Thiessen <opensource@fthiessen.de>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

export type { BaseOptions } from './baseConfig.js'
export type { AppOptions } from './appConfig'
export type { LibraryOptions } from './libConfig.js'

export { createBaseConfig } from './baseConfig.js'
export { createAppConfig } from './appConfig.js'
export { createLibConfig } from './libConfig.js'
