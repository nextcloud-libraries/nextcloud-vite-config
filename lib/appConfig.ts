/**
 * SPDX-FileCopyrightText: 2023 Ferdinand Thiessen <opensource@fthiessen.de>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type { Plugin, UserConfig, UserConfigFn } from 'vite'
import type { BaseOptions, NodePolyfillsOptions } from './baseConfig.js'

import { readFileSync } from 'node:fs'
import { relative } from 'node:path'
import { cwd } from 'node:process'
import { mergeConfig } from 'vite'
import { createBaseConfig } from './baseConfig.js'
import { findAppinfo } from './utils/appinfo.js'

import EmptyJSDirPlugin, { EmptyJSDirPluginOptions } from './plugins/EmptyJSDir.js'
import replace from '@rollup/plugin-replace'
import injectCSSPlugin from 'vite-plugin-css-injected-by-js'
import { CSSEntryPointsPlugin } from './plugins/CSSEntryPoints.js'

type VitePluginInjectCSSOptions = Parameters<typeof injectCSSPlugin>[0]

export interface AppOptions extends Omit<BaseOptions, 'inlineCSS'> {
	/**
	 * Override the `appName`, by default the name from the `appinfo/info.xml` and if not found the name from `package.json` is used.
	 * But if that name differs from the app id used for the Nextcloud app you need to override it.
	 */
	appName?: string

	/**
	 * Prefix to use for assets and chunks
	 * @default '{appName}-'
	 */
	assetsPrefix?: string

	/**
	 * Inject all styles inside the javascript bundle instead of emitting a .css file
	 * @default false
	 */
	inlineCSS?: boolean | VitePluginInjectCSSOptions,

	/**
	 * When not using inline css and using `cssCodeSplit` this option allows to create
	 * one CSS entry file for each JS entry point instead of only one for each JS entry point with styles.
	 *
	 * @default false
	 */
	createEmptyCSSEntryPoints?: boolean

	/**
	 * Whether to empty the 'js' directory
	 * Pass `false` to disable clearing the directory,
	 * it is also possible to pass options to the plugin.
	 * @default true
	 */
	emptyOutputDirectory?: boolean | EmptyJSDirPluginOptions

	/**
	 * Inject polyfills for node packages
	 * By default all node core modules are polyfilled, including prefixed with `node:` protocol
	 *
	 * @default '{ protocolImports: true }'
	 */
	nodePolyfills?: boolean | NodePolyfillsOptions

	/**
	 * Location of license summary file of third party dependencies
	 * Pass `false` to disable generating a license file.
	 *
	 * @default 'js/vendor.LICENSE.txt'
	 */
	thirdPartyLicense?: false | string
}

/**
 * Create a vite config for your nextcloud app
 *
 * @param entries Entry points of your app
 * @param options App related options for the vite config
 * @return {UserConfigFn} The vite config
 * @example
 * export default createAppConfig({
 *   main: path.resolve(path.join('src', 'main.js')),
 *   settings: path.resolve(path.join('src', 'settings.js')),
 * })
 */
export const createAppConfig = (entries: { [entryAlias: string]: string }, options: AppOptions = {}): UserConfigFn => {
	// Add default options
	options = {
		config: {},
		nodePolyfills: {
			protocolImports: true,
		},
		thirdPartyLicense: options.thirdPartyLicense === undefined ? 'js/vendor.LICENSE.txt' : options.thirdPartyLicense,
		...options,
	}

	let appVersion: string

	const appinfo = findAppinfo(cwd())
	if (appinfo) {
		const content = String(readFileSync(appinfo))
		const version = content.match(/<version>([^<]+)<\/version>/i)[1]
		const id = content.match(/<id>([^<]+)<\/id>/i)[1]

		if (version) {
			appVersion = version
		}
		if (id && !options.appName) {
			options.appName = id
		}
	} else {
		appVersion = process.env.npm_package_version
	}

	if (!options.appName) {
		console.warn('No app name configured, falling back to name from `package.json`')
		options.appName = process.env.npm_package_name
	}

	return createBaseConfig({
		...(options as BaseOptions),
		config: async (env) => {
			console.info(`Building ${options.appName} for ${env.mode}`)

			const assetsPrefix = (options.assetsPrefix ?? `${options.appName}-`).replace(/[/\\]/, '-')

			// This config is used to extend or override our base config
			// Make sure we get a user config and not a promise or a user config function
			const userConfig = await Promise.resolve(typeof options.config === 'function' ? options.config(env) : options.config)

			const plugins = [] as Plugin[]
			// Inject all imported styles into the javascript bundle by creating dynamic styles on the document
			if (options.inlineCSS) {
				const plugin = injectCSSPlugin({
					dev: {
						enableDev: env.mode === 'development',
					},
					...(typeof options.inlineCSS === 'object' ? options.inlineCSS : {}),
				})
				plugins.push(...[plugin].flat())
			} else if (userConfig.build?.cssCodeSplit) {
				// If not inlining CSS and using `cssCodeSplit` we need this plugin to fix https://github.com/vitejs/vite/issues/17527
				plugins.push(CSSEntryPointsPlugin({ createEmptyEntryPoints: options.createEmptyCSSEntryPoints }))
			}

			// defaults to true so only not adding if explicitly set to false
			if (options?.emptyOutputDirectory !== false) {
				// Ensure `js/` is empty as we can not use the build in option (see below)
				plugins.push(
					EmptyJSDirPlugin(
						typeof options.emptyOutputDirectory === 'object'
							? options.emptyOutputDirectory
							: undefined
					),
				)
			}

			// When building in serve mode (e.g. unit tests with vite) the intro option below will be ignored, so we must replace that values
			if (env.command === 'serve') {
				plugins.push(replace({
					delimiters: ['\\b', '\\b'],
					include: ['src/**'],
					preventAssignment: true,
					values: {
						appName: JSON.stringify(options.appName),
						appVersion: JSON.stringify(appVersion),
					},
				}))
			}

			// Extended config for apps
			const overrides: UserConfig = {
				plugins,
				// we need to set the base path, so module preloading works correctly
				// currently this is hidden behind the `experimental` options.
				experimental: {
					renderBuiltUrl(filename, { hostType }) {
						if (hostType === 'css') {
							// CSS `url()` does not support any dynamic path, so we use relative path to the css files
							return relative('../css', `../${filename}`)
						}
						return {
							// already contains the "js/" prefix as it is our output file configuration
							runtime: `window.OC.filePath('${options.appName}', '', '${filename}')`,
						}
					},
				},
				build: {
					// Output dir is the project root to allow main style to be generated within `/css`
					outDir: '',
					emptyOutDir: false, // ensure project root is NOT emptied!
					rollupOptions: {
						input: {
							...entries,
						},
						output: {
							// global variables for appName and appVersion
							intro: `const appName = ${JSON.stringify(options.appName)}; const appVersion = ${JSON.stringify(appVersion)};`,
							assetFileNames: (assetInfo) => {
								// Allow to customize the asset file names
								if (options.assetFileNames) {
									const customName = options.assetFileNames(assetInfo)
									if (customName) {
										return customName
									}
								}

								const extType = assetInfo.name.split('.').pop()
								if (/png|jpe?g|svg|gif|tiff|bmp|ico/i.test(extType)) {
									return 'img/[name][extname]'
								} else if (/css/i.test(extType)) {
									return `css/${assetsPrefix}[name].css`
								} else if (/woff2?|ttf|otf/i.test(extType)) {
									return 'css/fonts/[name][extname]'
								}
								return 'dist/[name]-[hash][extname]'
							},
							entryFileNames: () => {
								return `js/${assetsPrefix}[name].mjs`
							},
							chunkFileNames: () => {
								return 'js/[name].chunk.mjs'
							},
							manualChunks: {
								...(options?.coreJS ? { polyfill: ['core-js'] } : {}),
							},
						},
					},
				},
			}

			return mergeConfig(overrides, userConfig)
		},
	})
}
