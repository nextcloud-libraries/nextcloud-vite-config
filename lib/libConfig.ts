/**
 * SPDX-FileCopyrightText: 2023 Ferdinand Thiessen <opensource@fthiessen.de>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type { LibraryFormats, UserConfig, UserConfigFn, BuildOptions, Plugin, Rollup } from 'vite'
import type { BaseOptions } from './baseConfig.js'

import { mergeConfig } from 'vite'
import { createBaseConfig } from './baseConfig.js'

import DTSPlugin, { type PluginOptions as DTSOptions } from 'vite-plugin-dts'
import { nodeExternals, type ExternalsOptions } from 'rollup-plugin-node-externals'
import { ImportCSSPlugin } from './plugins/ImportCSS.js'

type OutputOptions = BuildOptions['rollupOptions']['output']

export interface LibraryOptions extends BaseOptions {
	/**
	 * Whether to minify the output
	 * @default false For libraries the code is not minified by default for better DX. Usually it is not needed: a library will be minified as a part of an app bundling.
	 */
	minify?: boolean

	/**
	 * Options for the rollup node externals plugin
	 *
	 * By default all `dependencies` and `peerDependencies` are marked as external.
	 * And node builtins prefix (`node:`) is stripped to make the library compatible with webpack and `node-polyfill-webpack-plugin`.
	 *
	 * Note: If you use dependencies `@nextcloud/vue/dist/Components/NcButton.js` and what them to be externalized too,
	 * you need to set an include pattern: `{ include: [ /^@nextcloud\/vue/ ]}`
	 */
	nodeExternalsOptions?: ExternalsOptions

	/**
	 * Options for the Vite DTS plugin
	 *
	 * This plugin allows to create .d.ts files for your library including the .vue files
	 * Pass `false` to disable the plugin
	 */
	DTSPluginOptions?: DTSOptions | false

	/**
	 * Formats you like your library to be built
	 * @default ['es']
	 */
	libraryFormats?: LibraryFormats[]

	/**
	 * Filename of the css output when a single CSS file should be extracted (`inlineCSS: false`)
	 * Note this only works with Vite 6.
	 *
	 * @default `style.css`
	 */
	cssFileName?: string
}

/**
 * Create a vite config for your nextcloud libraries
 *
 * @param entries Entry points of your app
 * @param options Options to use
 * @return {UserConfigFn} The vite config
 */
export const createLibConfig = (entries: { [entryAlias: string]: string }, options: LibraryOptions = {}): UserConfigFn => {
	// Add default values for options
	options = {
		config: {},
		minify: false,
		nodeExternalsOptions: {},
		libraryFormats: ['es'],
		cssFileName: 'style',
		...options,
	}

	return createBaseConfig({
		...options,
		config: async (env) => {
			const node = await nodeExternals({
				builtins: true, // Mark all node core modules, like `path` as external
				builtinsPrefix: 'strip', // Strip off `node:` prefix to make packages work with webpack (webpack/webpack#14166 and Richienb/node-polyfill-webpack-plugin#19)
				devDeps: false, // Development dependencies should not be included in the bundle
				peerDeps: true, // Peer dependencies should be by definition external
				deps: true, // Runtime dependencies: Same as with peer dependencies
				...options.nodeExternalsOptions,
			}) as Plugin

			// Order is important, run the plugin first
			node.enforce = 'pre'
			const plugins = [
				// Make dependencies external
				node,
			]

			// Handle inline CSS, this is different on library than on app mode, because app will have a DOM env so styles can injected in the document while libraries might run in node
			if (options.inlineCSS) {
				plugins.push(ImportCSSPlugin())
			}

			// Handle the DTS plugin
			if (options?.DTSPluginOptions !== false) {
				plugins.push(DTSPlugin(options.DTSPluginOptions))
			}

			// This config is used to extend or override our base config
			// Make sure we get a user config and not a promise or a user config function
			const userConfig = await Promise.resolve(typeof options.config === 'function' ? options.config(env) : options.config)

			const assetFileNames = (assetInfo: Rollup.PreRenderedAsset) => {
				// Allow to customize the asset file names
				if (options.assetFileNames) {
					const customName = options.assetFileNames(assetInfo)
					if (customName) {
						return customName
					}
				}

				const [name] = assetInfo.names
				const extType = name.split('.').pop()
				if (!options.inlineCSS && /css/i.test(extType)) {
					return '[name].css'
				}
				return 'assets/[name]-[hash][extname]'
			}

			// Manually define output options for file extensions
			const outputOptions: OutputOptions = options.libraryFormats.map(format => {
				const extension = format === 'es' ? 'mjs' : (format === 'cjs' ? 'cjs' : `${format}.js`)
				return {
					format,
					interop: 'auto', // Add __esModule for CJS externals imports to fix interop issues in tools like Babel/TS
					hoistTransitiveImports: false, // For libraries this might otherwise introduce side effects
					preserveModules: false,
					assetFileNames,
					entryFileNames: `[name].${extension}`,
					chunkFileNames: `chunks/[name]-[hash].${extension}`,
				}
			})

			return mergeConfig({
				plugins,
				build: {
					lib: {
						entry: {
							...entries,
						},
						formats: options.libraryFormats,
						cssFileName: options.cssFileName,
					},
					cssCodeSplit: true,
					outDir: 'dist',
					rollupOptions: {
						external: [/^core-js\//],
						output: outputOptions,
					},
				},
			} as UserConfig, userConfig)
		},
	})
}
