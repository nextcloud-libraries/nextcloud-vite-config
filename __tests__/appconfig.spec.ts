/**
 * SPDX-FileCopyrightText: 2023 Ferdinand Thiessen <opensource@fthiessen.de>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

// ok as this is just for tests
// eslint-disable-next-line n/no-extraneous-import
import type { RollupOutput, OutputOptions, OutputChunk } from 'rollup'
import { build, resolveConfig } from 'vite'
import { describe, it, expect } from 'vitest'
import { AppOptions, createAppConfig } from '../lib/appConfig'
import { fileURLToPath } from 'url'
import { resolve } from 'path'
import cssInjectedByJsPlugin from 'vite-plugin-css-injected-by-js'

const __dirname = fileURLToPath(new URL('.', import.meta.url))

describe('app config', () => {
	it('adds global appName variable', async () => {
		const resolved = await createConfig('build', 'development')
		expect([resolved.build.rollupOptions.output!].flat()[0].intro).toMatch(/appName = "@nextcloud\/vite-config"/)
	})

	it('adds global appName variable with override', async () => {
		const resolved = await createConfig('build', 'development', { appName: 'spreed' })
		expect([resolved.build.rollupOptions.output!].flat()[0].intro).toMatch(/appName = "spreed"/)
	})

	it('replaces process.env', async () => {
		const root = resolve(__dirname, '../__fixtures__/app_process_env')

		const resolved = await createAppConfig({
			main: resolve(root, 'main.js'),
		})({ command: 'build', mode: 'production' })

		const result = await build({
			...resolved,
			root,
		})
		const { output } = result as RollupOutput
		const code = (output[0] as OutputChunk).code
		expect(code.includes('process.env')).toBe(false)
		expect(code).toMatchSnapshot()
	})

	it('moves CSS assets to css/', async () => {
		const resolved = await createConfig('build', 'development')

		const output = resolved.build.rollupOptions.output as OutputOptions
		expect(typeof output?.assetFileNames).toBe('function')
		const assetFileNames = output?.assetFileNames as ((chunkInfo: unknown) => string)
		expect(assetFileNames({ names: ['some.css'] })).toMatch(/^css\/[^/]+\.css/)
		expect(assetFileNames({ names: ['other/file.css'] })).toMatch(/^css\/[^/]+\.css/)
	})

	it('moves image assets to img/', async () => {
		const resolved = await createConfig('build', 'development')

		const output = resolved.build.rollupOptions.output as OutputOptions
		expect(typeof output?.assetFileNames).toBe('function')
		const assetFileNames = output?.assetFileNames as ((chunkInfo: unknown) => string)
		expect(assetFileNames({ names: ['some.png'] })).toBe('img/[name][extname]')
		expect(assetFileNames({ names: ['some.svg'] })).toBe('img/[name][extname]')
		expect(assetFileNames({ names: ['some.jpg'] })).toBe('img/[name][extname]')
		expect(assetFileNames({ names: ['some.ico'] })).toBe('img/[name][extname]')
	})

	it('moves fonts to css/fonts', async () => {
		const resolved = await createConfig('build', 'development')

		const output = resolved.build.rollupOptions.output as OutputOptions
		expect(typeof output?.assetFileNames).toBe('function')
		const assetFileNames = output?.assetFileNames as ((chunkInfo: unknown) => string)
		expect(assetFileNames({ names: ['some.woff'] })).toBe('css/fonts/[name][extname]')
		expect(assetFileNames({ names: ['some.woff2'] })).toBe('css/fonts/[name][extname]')
		expect(assetFileNames({ names: ['some.otf'] })).toBe('css/fonts/[name][extname]')
		expect(assetFileNames({ names: ['some.ttf'] })).toBe('css/fonts/[name][extname]')
	})

	it('allow custom asset names', async () => {
		const resolved = await createConfig('build', 'development', { assetFileNames: ({ names }) => names[0] === 'main.png' ? 'img/main.png' : undefined })

		const output = resolved.build.rollupOptions.output as OutputOptions
		expect(typeof output?.assetFileNames).toBe('function')
		const assetFileNames = output?.assetFileNames as ((chunkInfo: unknown) => string)
		expect(assetFileNames({ names: ['main.png'] })).toBe('img/main.png')
		expect(assetFileNames({ names: ['foo.png'] })).toBe('img/[name][extname]')
	})

	it('extracts CSS by default with CSS entry points', async () => {
		const resolved = await createConfig('build', 'development')

		expect(resolved.build.cssCodeSplit).toBe(true)
		expect(resolved.plugins.find(({ name }) => name === 'css-entry-points-plugin')).not.toBeUndefined()
	})

	it('does not add CSS entry points with CSS inject', async () => {
		const resolved = await createConfig('build', 'development', { inlineCSS: true })

		expect(resolved.build.cssCodeSplit).toBe(true)
		// not found!
		expect(resolved.plugins.find(({ name }) => name === 'css-entry-points-plugin')).toBeUndefined()
	})

	describe('inlining css', () => {
		const pluginName = [cssInjectedByJsPlugin()].flat()[0].name

		it('does not inline css by default', async () => {
			const resolved = await createConfig()
			expect(resolved.plugins.filter(({ name }) => name === pluginName)).toHaveLength(0)
		})

		it('does not inline css when disabled', async () => {
			const resolved = await createConfig('build', 'production', { inlineCSS: false })
			expect(resolved.plugins.filter(({ name }) => name === pluginName)).toHaveLength(0)
		})

		it('does inline css when enabled', async () => {
			const resolved = await createConfig('build', 'production', { inlineCSS: true })
			expect(resolved.plugins.filter(({ name }) => name === pluginName)).toHaveLength(1)
		})

		it('does inline css when enabled with configuration', async () => {
			const resolved = await createConfig('build', 'production', { inlineCSS: { useStrictCSP: true } })
			expect(resolved.plugins.filter(({ name }) => name === pluginName)).toHaveLength(1)
		})
	})

	const createConfig = async (command: 'build' | 'serve' = 'build', mode: 'development' | 'production' = 'production', options?: AppOptions) => await resolveConfig(await createAppConfig({
		main: 'src/main.js',
	}, options)({ command, mode, isSsrBuild: false }), command)
})
