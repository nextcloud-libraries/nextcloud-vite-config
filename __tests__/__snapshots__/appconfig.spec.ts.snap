// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app config > replaces process.env 1`] = `
"/*! third party licenses: js/vendor.LICENSE.txt */
var e={};window.my_timezone=e.TZ;
"
`;
